File: labs/08-ocr/python/read-text/document_processing_service.py
import logging
import os
from azure.core.credentials import AzureKeyCredential
from azure.ai.documentintelligence import DocumentIntelligenceClient
from azure.core.exceptions import HttpResponseError

# Configure logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

class DocumentProcessingService:
    def __init__(self, endpoint, key):
        self.client = DocumentIntelligenceClient(endpoint=endpoint, credential=AzureKeyCredential(key))

    def process_documents(self, folder_path):
        document_files = [os.path.join(folder_path, f) for f in os.listdir(folder_path)
                          if os.path.isfile(os.path.join(folder_path, f)) and f.lower().endswith('.pdf')]

        if not document_files:
            logger.info(f"No PDF files found in {folder_path}.")
            return

        for document_file in document_files:
            self.read_document(document_file)

    def read_document(self, document_file_path):
        logger.info("----------------------------------------------------------")
        logger.info(f"READ FILE FROM LOCAL: {document_file_path}")
        logger.info("----------------------------------------------------------")

        try:
            with open(document_file_path, "rb") as document_stream:
                poller = self.client.begin_analyze_document(
                    "prebuilt-read",
                    analyze_request=document_stream,
                    content_type="application/octet-stream"
                )
                result = poller.result()

            # Detect languages.
            logger.info("\r\n----Languages detected in the document----\r\n")
            if result.languages is not None:
                for language in result.languages:
                    logger.info(f"Language code: '{language.locale}' with confidence {language.confidence}")

            # Analyze pages.
            for page in result.pages:
                logger.info(f"----Analyzing document from page #{page.page_number}----")
                logger.info(f"Page has width: {page.width} and height: {page.height}, measured with unit: {page.unit}")

                # Analyze lines.
                if page.lines:
                    for line in page.lines:                        
                        logger.info(f"'{line.content}'")
          

            logger.info("----------------------------------------------------------\r\n\r\n")

        except HttpResponseError as error:
            if error.error is not None:
                if error.error.code == "InvalidImage":
                    logger.error(f"Received an invalid image error: {error.error}")
                if error.error.code == "InvalidRequest":
                    logger.error(f"Received an invalid request error: {error.error}")
                raise
            if "Invalid request".casefold() in error.message.casefold():
                logger.error(f"Uh-oh! Seems there was an invalid request: {error}")
            raise
        except Exception as e:
            logger.error("An error occurred: %s", e)

File: labs/08-ocr/csharp/read-text/DocumentProcessingService.cs
using System;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using Azure;
using Azure.AI.DocumentIntelligence;
using Microsoft.Extensions.Configuration;

namespace Ai102.ReadText
{
    public class DocumentProcessingService
    {
        private readonly DocumentIntelligenceClient _client;

        public DocumentProcessingService(string endpoint, string key)
        {
            _client = new DocumentIntelligenceClient(new Uri(endpoint), new AzureKeyCredential(key));
        }

        public async Task ProcessDocumentsAsync(string folderPath)
        {
            string[] documentFiles = Directory.GetFiles(folderPath, "*.pdf", SearchOption.AllDirectories);

            if (documentFiles.Length == 0)
            {
                Console.WriteLine($"No PDF files found in {folderPath}.");
                return;
            }

            foreach (var documentFile in documentFiles)
            {
                await ReadFileAsync(documentFile);
            }
        }

        private async Task ReadFileAsync(string documentFilePath)
        {
            Console.WriteLine("----------------------------------------------------------");
            Console.WriteLine($"READ FILE FROM LOCAL: {documentFilePath}");
            Console.WriteLine("----------------------------------------------------------");


            using (FileStream stream = new FileStream(documentFilePath, FileMode.Open))
            {
                var binaryData = BinaryData.FromStream(stream);
                var content = new AnalyzeDocumentContent { Base64Source = binaryData };
                var operation = await _client.AnalyzeDocumentAsync(WaitUntil.Completed, "prebuilt-read", content);
                AnalyzeResult result = operation.Value;

                foreach (DocumentPage page in result.Pages)
                {
                    for (int i = 0; i < page.Lines.Count; i++)
                    {
                        DocumentLine line = page.Lines[i];
                        Console.WriteLine($"'{line.Content}'");
                    }
                }

                foreach (DocumentLanguage language in result.Languages)
                {
                    Console.WriteLine($"  Found language '{language.Locale}' with confidence {language.Confidence}.");
                }

                // Extract key-value pairs
                if (result.KeyValuePairs.Any())
                {
                    Console.WriteLine("Key-Value Pairs:");
                    foreach (var kvp in result.KeyValuePairs)
                    {
                        string key = kvp.Key.Content;
                        string value = kvp.Value.Content;
                        Console.WriteLine($"  {key}: {value}");
                    }
                }
                Console.WriteLine("----------------------------------------------------------\r\n\r\n");
            }

            Console.WriteLine("To learn more about Azure Document Intelligence features and capabilities yo can start with " +
            "https://learn.microsoft.com/en-us/samples/azure-samples/document-intelligence-code-samples/document-intelligence-code-samples/");
        }
    }
}
